src/libraries/System.Private.CoreLib/src/System/Delegate.cs

```diff
@@ -57,6 +57,67 @@ public abstract partial class Delegate : ICloneable, ISerializable
 
         public virtual Delegate[] GetInvocationList() => new Delegate[] { this };
 
+        /// <summary>
+        /// Gets a value that indicates whether the <see cref="Delegate"/> has a single invocation target.
+        /// </summary>
+        /// <value>true if the <see cref="Delegate"/> has a single invocation target.</value>
+        public bool HasSingleTarget => Unsafe.As<MulticastDelegate>(this).HasSingleTarget;
+
+        /// <summary>
+        /// Gets an enumerator for the invocation targets of this delegate.
+        /// </summary>
+        /// <remarks>
+        /// This returns a <see cref="InvocationListEnumerator{TDelegate}"/>" /> that follows the IEnumerable pattern and
+        /// thus can be used in a C# 'foreach' statements to retrieve the invocation targets of this delegate without allocations.
+        /// The order of the delegates returned by the enumerator is the same order in which the current delegate invokes the methods that those delegates represent.
+        /// </remarks>
+        public static System.Delegate.InvocationListEnumerator<TDelegate> EnumerateInvocationList<TDelegate>(TDelegate d) where TDelegate : System.Delegate
```

Comment:
@stephentoub Do you have an opinion about `null` handling? Return empty enumerator or throw NullReferenceException?

Reply:
You mean if someone passes in a null `d`? I think we should throw an ArgumentNullException if it's null.

Or do you mean if someone uses `default(InvocationListEnumerator)`? I'd let it null ref, as we do in `List<T>.Enumerator`.

Reply:
Err, actually, I take that back. It'd be pretty reasonable to use this with events, in which case the delegate would be null.  So, let's go with empty for null.

Reply:
(And given that, I think the parameter should be `TDelegate? d`, highlighting that nulls are permitted)